transformations/transformations/block_index_transformations.py

```diff
@@ -0,0 +1,428 @@
+        SCCBaseTransformation.resolve_masked_stmts(routine, loop_variable=v_index)
+
+        # Bail if routine is marked as sequential or routine has already been processed
+        if SCCBaseTransformation.check_routine_pragmas(routine, directive=None):
+            return
+
+        bounds = SCCBaseTransformation.get_horizontal_loop_bounds(routine, self.horizontal)
+        SCCBaseTransformation.resolve_vector_dimension(routine, loop_variable=v_index, bounds=bounds)
+
+        # for kernels we process the entire body
+        routine.body = self.process_body(routine.body, routine.symbol_map, item.trafo_data[self._key]['definitions'],
+                                         successors, targets, exclude_arrays)
+
+
+class BlockIndexInjectTransformation(Transformation):
+    """
+    A transformation pass to inject the block-index in arrays promoted by a previous transformation pass. As such,
+    this transformation also relies on the block-index, or a known alias, being *already* present in routines that
+    are to be transformed.
+
+    For array access in a :any:`Subroutine` body, it operates by comparing the local shape of an array with its
+    declared shape. If the local shape is of rank one less than the declared shape, then the block-index is appended
+    to the array's dimensions.
+
+    For :any:`CallStatement` arguments, if the rank of the argument is one less than that of the corresponding
+    dummy-argument, the block-index is appended to the argument's dimensions. It should be noted that this logic relies
+    on the :any:`CallStatement` being free of any sequence-association.
+
+    For example, the following code:
+
+    .. code-block:: fortran
+
+        subroutine kernel1(nblks, ...)
+           ...
+           integer, intent(in) :: nblks
+           integer :: ibl
+           real :: var(jlon,nlev,nblks)
+
+           do ibl=1,nblks
+             do jlon=1,nproma
+               var(jlon,:) = 0.
+             enddo
+
+             call kernel2(var,...)
+           enddo
+           ...
+        end subroutine kernel1
+
+        subroutine kernel2(var, ...)
+           ...
+           real :: var(jlon,nlev)
+        end subroutine kernel2
+
+    is transformed to:
+
+    .. code-block:: fortran
+
+        subroutine kernel1(nblks, ...)
+           ...
+           integer, intent(in) :: nblks
+           integer :: ibl
+           real :: var(jlon,nlev,nblks)
+
+           do ibl=1,nblks
+             do jlon=1,nproma
+               var(jlon,:,ibl) = 0.
+             enddo
+
+             call kernel2(var(:,:,ibl),...)
+           enddo
+           ...
+        end subroutine kernel1
+
+        subroutine kernel2(var, ...)
+           ...
+           real :: var(jlon,nlev)
+        end subroutine kernel2
+
+    Specific arrays in individual routines can also be marked for exclusion from this transformation by assigning
+    them to the `exclude_arrays` list in the :any:`SchedulerConfig`.
```

Comment:
This is to catch instances of sequence association, which will return false positives for the arg-dummy shape checks. Sequence associations are mostly disallowed by the new coding standards, so I think we can keep the simple shape-based logic here and have bail-outs for the few known offenders rather than try and resolve the sequence association (which would probably need more complex size-based checks).